disk_objectstore/utils.py

```diff
@@ -1519,3 +1534,117 @@ def merge_sorted(iterator1: Iterable[Any], iterator2: Iterable[Any]) -> Iterator
+        # Use the same compression type
+        return source_compressed
+    if compress_mode == CompressMode.AUTO:
+        if source_compressed:
+            # Only compress if it's worth it
+            return source_length / source_size < compression_threshold
+
+        # As noted by @dev-zero in #14:
+        #  > modern compression algorithms already automatically store uncompressed data if data is uncompressible
+        #  > (see the various tuning knobs of LZMA/LZ4/XZ), if possible use one of those as they usually
+        #  > already provide multithreaded implementations.
+        # Therefore, in the future if we support those clever algorithms we can just skip this and let the algorithm
+        # do its job.
+
+        # Estimate the amount of compression we can get and decide whether to compress based on how much we can save.
+        compression_ratio = estimate_compression(source_stream, source_size)
+        return compression_ratio < compression_threshold
+    raise NotImplementedError(f"Unknown {compress_mode=}")
+
+
+def estimate_compression(stream: StreamSeekBytesType, size: int) -> float:
+    """Tries to quickly estimate the compression ratio of the file.
+
+    At the end the stream, put back at the same position as when we entered this method (i.e. we will seek back
+    at the stream.tell() position that was set at the function enter).
+
+    :param stream: a seekable stream.
+    :param size: the total size of the stream (this is used to try to get small chunks of data from small
+        random parts of the stream.
+    :return: the ratio between the expected compressed size and the total file size. E.g. 1. means compressing
+        does not provide any feedback, 0.5 means compression halves the files size, 0.1 means the compressed version
+        only occupies 10% of the space, a value larger than 1 means that compressing actually increases the file size,
+        etc. Note that this is only a quick estimate; actual compression levels might vary, but it is meant as a
+        quick heuristics to decide whether we should compress or not, when CompressMode.AUTO is requested.
+    """
+    # Set some internal parameters
+    sample_size = 1024  # Get this amount of bytes...
+    # Only check a sampled_data of at most `max_sample_data_size` bytes (approximately)
+    max_sampled_data_size = 131_072  # 128kB
+    # Approximately, except for rounding errors, get a small sample of size `sample_size` every so many bytes
+    sample_interval = size // (max_sampled_data_size // sample_size)
+
+    # Get the current pos (to go back to the same position at the end)
+    initial_pos = stream.tell()
+    # Go back to the beginning
+    stream.seek(0)
+
+    sampled_data = b""
+    # Only check the first part of the file
+    while len(sampled_data) < max_sampled_data_size:
+        chunk = stream.read(sample_size)
+        if not chunk:
+            # EOF, stop
+            break
+        sampled_data += chunk
```

Comment:
A minor point - this would create a copy each time, and the loop is called up to 128 times. Probably better to use `BytesIO` instead?

Reply:
OK - I will replace with a list of bytes, that I join only at the end.